### PR TITLE
Add option to force resumable uploads.

### PR DESCRIPTION
**Pull request by coryan**
The library automatically selects resumable uploads when it can
determine that this is the best upload strategy. But sometimes the
application simply knows that a resumable upload is necessary, in this
case the application developer should be able to force a resumable
upload. In addition, setting this option to a non-empty string will be
used to restore a previously created resumable upload.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1566)

### CODE REVIEW

File: google/cloud/storage/upload_options.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * Request a resumable upload, restoring a previous session if necessary.
+ *
+ * When this option is used the client library prefers using resumable uploads.
+ *
+ * If the value passed to this option is the empty string, then the library will
+ * create a new resumable session. Otherwise the value should be the id of a
+ * previous upload sesion, the client library will restore that session in
+ * this case.
+ */
+struct UseResumableUploadSession
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/upload_options.h]: Good point.  Unfortunately the default constructors for all options must initialize the option to "not set", but what about creating a `gcs::NewResumableUploadSession()` and `gcs::RestoreResumableUploadSession(std::string)` functions that return the same type?

**Comment by coryan** [comment]
PTAL.

**Comment by coryan** [comment]
Rebased to pick up the Windows fixes.

**Comment by coryan** [comment]
Ping.

File: google/cloud/storage/upload_options.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (houglum) [reply google/cloud/storage/upload_options.h]: That seems fine as well -- just looking for something to clearly disambiguate between the two scenarios for users :)

[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/upload_options.h]: Done.

**Pull request merged**